plugins/bundle/plugin.go

```diff
@@ -517,6 +518,18 @@ func (p *Plugin) process(ctx context.Context, name string, u download.Update) {
 				}
 				return
 			}
+
+			err = p.saveBundleManifestToDisk(name, u.ETag)
+			if err != nil {
+				p.log(name).Error("Persisting bundle manifest to disk failed: %v", err)
+				p.status[name].SetError(err)
+				if !p.stopped {
+					etag := p.etags[name]
+					p.downloaders[name].SetCache(etag)
+				}
+				return
+			}
+
```

Comment:
There could an error condition here I think. So let's say we already have a persisted bundle and manifest file. Now OPA attempts to persist a new bundle and that attempt fails but `saveBundleToDisk` returns no error as there is a bundle already persisted. Now say `saveBundleManifestToDisk` is successful and the manifest file is updated. So now you have a mismatch while loading the persisted bundle. To avoid this, let's move `saveBundleManifestToDisk` as part of `saveBundleToDisk` and use the rename trick in the `saveBundleToDisk` for both the bundle and manifest.

But I think even if we do that there is still the possibility that the bundle and the manifest are a mismatch.

Reply:
> use the rename trick in the saveBundleToDisk for both the bundle and manifest

I can have a look into doing this 👍 and see what I can do in addition to avoid chances of mismatches.

Reply:
I have added this in 8fa37fbb1. I think that using the rename trick is reasonably defensive. The only failure case I can think of with the current implementation is:

1. Bundle is downloaded with ETag foo and saved in temp file
1. Manifest is created with ETag foo and saved in temp manifest file
1. Bundle file is renamed to correct path, operation completes ok
2. Manifest file fails to move to correct location
3. OPA throws error, operation will be retried

If that happens, either:

* A) an old manifest file will be there, or
* B) there will be no manifest file.

Given that the ETag will still be in memory anyway this is unlikely to be an issue in a running OPA instance.

If OPA is restarted however, A) is undesirable and worth avoiding. B) is acceptable and not worth avoiding.

One way we can avoid A) is by removing the manifest file before renaming the bundle file. With this approach we have 4 failure cases:

* fail to remove old manifest file. This is ok since there are no changes to state.
* fail to move bundle. This is ok, old bundle remains, manifest file is missing now so surplus download may happen.
* fail to move manifest file. This is ok, new bundle is in place, manifest file is missing so surplus download may happen.

This is what I have added in 845b677c71f37758cec151c076a5666715773957.

Reply:
Thanks @charlieegan3 for looking into this. 

This seems to me that there is still possibility of errors happening here given these cases and eventually we anyways end up downloading the bundle on start-up. So it's not clear to me what benefits the additional complexity gives us. Is there a way to simplify this? If we had both the bundle and etag as part of one struct, this would make things simpler I think. For ex, 

```go
type bundlePackage struct {
    RawBundle []byte
    Etag string 
}
```
We can then pass this to the persist and retrieve logic. I haven't tried this so not sure if this works or is feasible.  But at a high level if there is a way to to keep raw bundle and etag as one unit, these ops will become much simpler.

Reply:
> So it's not clear to me what benefits the additional complexity gives us

Yeah, perhaps it isn't worth it. It does make the failures less likely, but perhaps not enough of an improvement.

I suppose that the reason I'd not gone down the monolith bundlePackage route is that it's not backwards compatible. I thought it was possible that users might be using the persisted bundle directory to kickstart OPA in some cases - perhaps this is not a use case we see as common enough to worry about though?

I'm not sure what the relationship between bundle Etag and manifest revision is expected to be in most cases. Would using the revision from a loaded bundle's manifest file as the value for `if-none-match` only add more confusion?

Reply:
> I'm not sure what the relationship between bundle Etag and manifest revision is expected to be in most cases. Would using the revision from a loaded bundle's manifest file as the value for if-none-match only add more confusion?

They could be same or different. It depends on the bundle creator. So we cannot assume the revision can be used  in the if-none-match header to replace the etag. 

> I suppose that the reason I'd not gone down the monolith bundlePackage route is that it's not backwards compatible. 

That's a good point. Although we say OPA will try to load and activate persisted bundles on a best-effort basis this could break some folks. I don't have an exact number on the impact of this change. This is another factor to take into account I think.

Reply:
Ok thanks Ash. I'll see if I can get together a revised implementation for this tomorrow. I'll aim to:

* still support the loading of bundle.tar.gz for older versions
* only have one write operation when saving
* serialize something like `bundlePackage` so that the data included in there can be easily extended if need be in the future.

Will report back once that's done. Thanks for the back and forth here! 🙏

Reply:
I've had another go at this. The main implementation is in `internal/bundle/utils.go`. let me know if you think that's like what you had in mind 🙂

Reply:
I had an idea that it might be possible to use the gzip comment field to store some extra information... I've added that in the latest commit.

https://github.com/open-policy-agent/opa/pull/5687/files#diff-b7bffa44a68a4b43193ab3157de6ff3844353f7b4903663a7552d1294dda1526R802

What's nice about this is what it means that we don't need to create a bundlePackage in memory and reuse most of the existing bundle write logic. What's less nice is that we're potentially abusing the gzip header comment field to store data. Let me know what you think.